lib/relooper/src/graph/cfg/mod.rs

```diff
@@ -0,0 +1,175 @@
+use crate::graph::cfg::CfgEdge::{Cond, Terminal, Uncond};
+use crate::traversal::graph::bfs::Bfs;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::Hash;
+
+mod cfg_parsing;
+
+pub trait CfgLabel: Copy + Hash + Eq + Ord + Sized {}
+
+impl<T: Copy + Hash + Eq + Ord + Sized> CfgLabel for T {}
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub enum CfgEdge<TLabel> {
+    Uncond(TLabel),
+    Cond(TLabel, TLabel),
+    Terminal,
+}
+
+impl<TLabel> CfgEdge<TLabel> {
+    pub fn to_vec(&self) -> Vec<&TLabel> {
+        match self {
+            Uncond(u) => vec![u],
+            Cond(cond, fallthrough) => vec![cond, fallthrough],
+            Terminal => vec![],
+        }
+    }
+
+    pub(crate) fn map<'a, U, F: Fn(&'a TLabel) -> U>(&'a self, mapping: F) -> CfgEdge<U> {
+        match self {
+            Uncond(t) => Uncond(mapping(t)),
+            Cond(t, f) => Cond(mapping(t), mapping(f)),
+            Terminal => Terminal,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Cfg<TLabel> {
+    pub(crate) entry: TLabel,
+    out_edges: HashMap<TLabel, CfgEdge<TLabel>>,
+}
+
+impl<T> Cfg<T> {
+    pub fn edges(&self) -> &HashMap<T, CfgEdge<T>> {
+        &self.out_edges
+    }
+}
+
+impl<T: Eq + Hash + Clone> Cfg<T> {
+    pub fn new(entry: T) -> Cfg<T> {
+        Self {
+            entry,
+            out_edges: Default::default(),
+        }
+    }
+
+    pub fn map_label<'a, M, U: Eq + Hash>(&'a self, mapping: M) -> Cfg<U>
+    where
+        M: Fn(&'a T) -> U,
+    {
+        let out_edges: HashMap<U, CfgEdge<U>> = self
+            .out_edges
+            .iter()
+            .map(|(from, e)| (mapping(from), e.map(&mapping)))
+            .collect();
+
+        Cfg {
+            entry: mapping(&self.entry),
+            out_edges,
+        }
+    }
+
+    pub fn to_borrowed(&self) -> Cfg<&T> {
+        self.map_label(|l| l)
+    }
+
+    pub fn nodes(&self) -> HashSet<&T> {
+        self.out_edges.keys().collect()
+    }
+
+    pub fn children(&self, label: &T) -> HashSet<&T> {
+        self.out_edges
+            .get(label)
+            .into_iter()
+            .flat_map(|edge| edge.to_vec())
+            .collect()
+    }
+
+    pub fn add_edge(&mut self, from: T, edge: CfgEdge<T>) {
+        let nodes = self.nodes();
+        // todo unncecessary collect
+        for n in edge
+            .to_vec()
+            .into_iter()
+            .filter(|n| !nodes.contains(n))
+            .collect::<Vec<_>>()
+        {
+            // todo clone
```

Comment:
Remaining todo

Reply:
I think there is no way to avoid this clone. You need insert ``n`` to ``self.nodes`` and later, in line 103, insert ``edge`` into ``self.outedges``.

Reply:
Unfortunately it's is not trivial problem with borrow checker, I think now is better just left it as is